### PR TITLE
feat(automation): new world doc pr creation

### PR DESCRIPTION
**Pull request by loicsay**
### [DEX-1021](https://algolia.atlassian.net/browse/DEX-1021)

This PR create a command to generate a `new-world-docs` PR that'll update the `.mdx` files for the CLI documentation.

## How to test?
 
```sh
go run ./cmd/docs --app_data-path docs/app_data/cli/commands --target new
```

It will output the files in at `docs/app_data/cli/commands`

## Refinement

Summary of our discussion with @clemfromspace:

- We'll add annotation (`runInWebCli`) to commands with `web CLI` enabled in `new-world-docs`
- Because there's differences between CLI examples and the ones run with the `web CLI`, we'll standardize all examples with the `new-world-docs`/`webCLI` (in a future PR)
- Examples with output or pipe symbols will automatically be non-runnable

[DEX-1021]: https://algolia.atlassian.net/browse/DEX-1021?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by loicsay** [comment]
@dylantientcheu @clemfromspace 
I'm not sure about how we should handle the `runInCommand` property for each example 🤔 
Do you have any idea? Should we hardcode a list of commands that allow this feature in the CLI codebase?

**Comment by dylantientcheu** [comment]
For now we can edit the submitted PR each time, i'd prefer not to bloat the cli code.

**Comment by clemfromspace** [comment]
I think we should run the example directly, so we will have to edit the example to make them match what indices / things we have in the doc. That way, no editing is necessary and it's more coherent maybe. WDYT?

**Comment by loicsay** [comment]
@clemfromspace I agree, let's remove all `Examples` formatting for now and do it in `new-world-doc` 👍 
Same for the WebCLI I think? You're suggesting doing outside of this repo if I understand correctly?

File: internal/docs/yaml.go [hunk 1]
```diff
@@ -18,6 +18,7 @@ type Command struct {
 	Usage       string
 	Aliases     []string
 	Examples    string
+	Slug        string
```
Comment (clemfromspace) [inline internal/docs/yaml.go:21]: Some of these things are shared between `yaml` and `mdx`, what do you think about creating a `docs.go` file for the shared structs & functions?

File: internal/docs/mdx.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+		}
+		command := newCommand(c)
+		command.Slug = strings.ReplaceAll(command.Name, " ", "-")
+		filename := filepath.Join(dir, fmt.Sprintf("%s.mdx", command.Slug))
+		file, err := os.Create(filename)
+		if err != nil {
+			return err
+		}
+
+		tpl, err := template.ParseFiles("internal/docs/mdx.tpl")
```
Comment (clemfromspace) [inline internal/docs/mdx.go:26]: This should be outside of the loop.

File: internal/docs/mdx.tpl [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+{{ end }}
+{{ else }}
+## {{ .Name }}
+### Usage
+
+`{{ .Usage }}`
+
+{{ if .Examples }}
+### Examples
+{{ .Examples }}
```
Comment (clemfromspace) [inline internal/docs/mdx.tpl:36]: In this case, the examples won't even be wrapped between backticks?
For the new docs, I would also split the examples so we can more easily copy them / run them in the webcli.
Adding a new method to the `Command` struct like that:
```go
type Example struct {
	Desc   string
	Code   string
	WebCLI string
}

func (cmd Command) ExamplesList() []Example {
	var examples []Example
	examplesRaw := strings.Split(cmd.Examples, "#")
	for _, example := range examplesRaw {
		example = strings.TrimSpace(example)
		if len(example) == 0 {
			continue
		}
		exampleLines := strings.Split(example, "\n")

		code := strings.ReplaceAll(exampleLines[1], "$", "")
		code = strings.TrimSpace(code)

		examples = append(examples, Example{
			Desc: strings.Replace(exampleLines[0], "#", "", 1),
			Code: code,
		})

		// TODO: Add WebCLI example depending of:
		// 1. The presence of the `runInWebCLI` annotation on the command
		// 2. The presence of any piping or file flags in the example
	}
	return examples
}
```

WDYT?

File: internal/docs/mdx.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/spf13/cobra"
+)
+
+func GenMdxFile(cmd *cobra.Command, dir string) error {
```
Comment (clemfromspace) [inline internal/docs/mdx.go:13]: Could we keep a consistent naming between the `yaml` and `mdx` version?
```suggestion
func GenMdxTree(cmd *cobra.Command, dir string) error {
```

File: internal/docs/mdx.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	for _, c := range cmd.Commands() {
```
Comment (clemfromspace) [inline internal/docs/mdx.go:14]: Could we build the tree of the commands at the same place between the two version? Many of those lines are duplicated.

Maybe in the `docs.go`, adding a `getCommands` like
```go
func getCommands(cmd *cobra.Command) []Command {
	var commands []Command
	for _, c := range cmd.Commands() {
		if c.Hidden || c.Name() == "help" {
			continue
		}
		command := newCommand(c)
		if c.HasAvailableSubCommands() {
			for _, sub := range c.Commands() {
				command.SubCommands = append(command.SubCommands, newCommand(sub))
			}
		}
		commands = append(commands, command)
	}

	return commands
}
```

**Review by clemfromspace — CHANGES_REQUESTED** [review]
Good job, the template way was definitely the way to go 👍 

I think we need to work more of the example presentation (splitting, etc.)
I also added some comments regarding the duplicated code / now shared code between the two docs generation methods.

**Comment by clemfromspace** [comment]
@loicsay I saw that you actually removed the example splitting, this was not what I wanted to say 😓 
Sorry about the confusion, I wanted to say that we should directly run the examples in the webCLI (for the available commands), but that we should edit those examples, directly in the CLI, to match the indices names & co that are present in the docs. This should be done maybe in a second step, along with adding a `availableInWebCLI` annotation on the commands.

**Comment by loicsay** [comment]
Summary of our discussion with @clemfromspace:

- We'll add annotation (`runInWebCli`) to commands with `web CLI` enabled in `new-world-docs`
- Because there's differences between CLI examples and the ones run with the `web CLI`, we'll standardize all examples with the `new-world-docs`/`webCLI` (in a future PR)
- Examples with output or pipe symbols will automatically be non-runnable

File: pkg/cmd/indices/list/list.go [hunk 6]
```diff
@@ -43,6 +43,9 @@ func NewListCmd(f *cmdutil.Factory) *cobra.Command {
 		RunE: func(cmd *cobra.Command, args []string) error {
 			return runListCmd(opts)
 		},
+		Annotations: map[string]string{
```
Comment (loicsay) [inline pkg/cmd/indices/list/list.go:46]: Only added this one and `search.go` for testing, will add it to all other commands that can be run with the web CLI with the second PR that'll take care of standardize examples

**Review by clemfromspace — APPROVED** [review]
Love it, LGTM 🚀

**Pull request merged**